### ISSUE TITLE
Cannot deploy Tensorflow 2.0.0 to Google cloud functions

### ISSUE THREAD

**Issue by m4recek** [tensorflow/tensorflow#34707]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Google cloud functions
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): binary
- TensorFlow version: 2.0.0
- Python version: 3.7
- Installed using virtualenv? pip? conda?: pip
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

We are getting following error while trying to deploy Tensorflow 2.0.0. 

```
gcloud functions deploy matchProbability --runtime python37 --trigger-http --memory 2048MB --timeout=540 --set-env-vars TEMP=/tmp/
```

```
Deploying function (may take a while - up to 2 minutes)...failed.                                                                                                                         
ERROR: (gcloud.functions.deploy) OperationError: code=3, message=Build failed: {"error": {"canonicalCode": "INVALID_ARGUMENT", "errorMessage": "'pip_download_wheels' had stderr output:\n 
 Could not find a version that satisfies the requirement tensorflow==2.0.0 (from -r requirements.txt (line 53)) (from versions: 1.13.0rc1, 1.13.0rc2, 1.13.1, 1.13.2, 1.14.0rc0, 1.14.0rc1,
 1.14.0, 2.0.0a0, 2.0.0b0, 2.0.0b1)\nNo matching distribution found for tensorflow==2.0.0 (from -r requirements.txt (line 53))\n\nerror: 'pip_download_wheels' returned code: 1", "errorTyp
e": "InternalError", "errorId": "72357014"}}
```

Not sure what are next steps should be, only [article](https://cloud.google.com/blog/products/ai-machine-learning/how-to-serve-deep-learning-models-using-tensorflow-2-0-with-cloud-functions) describing deploying Tensorflow to google cloud functions is using version 2.0.0b1. Should we downgrade?

**Comment by av8ramit** [comment]
I know we changed TensorFlow 2.0 official to be [manylinux2010](https://www.python.org/dev/peps/pep-0571/) compliant between beta and official which requires an updated pip. Could that potentially be the issue?

There is definitely a [Python3.7 binary](https://pypi.org/project/tensorflow/#files) for tensorflow==2.0.0.

**Comment by ymodak** [comment]
Required pip version is 19.0 or later for TF 2.0.0
See https://www.tensorflow.org/install/pip?lang=python3#system-requirements

**Comment by m4recek** [comment]
Hi, I can confirm this is problem with Google Cloud Function. That article is just wrong and should not exists...

They are still running on PIP 18.0, thats why Tensorflow 2.0.0 (and any newer software) cannot be installed. Obviously, same will be true for TF 2.1.0 when it will be released.

Whats worse according to private support they have no plan to update and marked it as "feature request", so it has no resolution time.

Here is public support ticket created for this issue, you can consider staring it: https://issuetracker.google.com/145374253

**Comment by northtree** [comment]
> Hi, I can confirm this is problem with Google Cloud Function. That article is just wrong and should not exists...
> 
> They are still running on PIP 18.0, thats why Tensorflow 2.0.0 (and any newer software) cannot be installed. Obviously, same will be true for TF 2.1.0 when it will be released.
> 
> Whats worse according to private support they have no plan to update and marked it as "feature request", so it has no resolution time.
> 
> Here is public support ticket created for this issue, you can consider staring it: https://issuetracker.google.com/145374253

At this moment, GCF supports latest version of pip, which is `20.0.2`.

You could verify via this GCF. 

```python
def pip_version(_):
  import pip
  print(pip.__version__)
```

Deploy and test. `gcloud functions deploy pip_version --runtime python37 --trigger-http`

Moreover, your statement about TF2 is also not true. From [GCP blog](https://cloud.google.com/blog/products/ai-machine-learning/how-to-serve-deep-learning-models-using-tensorflow-2-0-with-cloud-functions), TF2 was supported to run via GCF since July 2019.

**Comment by bonaventura-p** [comment]
@northtree You are correct. However GCF fails to install any version of TF2 > 2.0.0b . I had been trying to deploy a GCF with TF 2.1.0 and it was failing. Downgrading to 2.0.0b seems to fix the issue for now. But any hints on what is preventing GCF from deploying newer versions of TF2?

**Comment by m4recek** [comment]
@northtree No, Google Cloud Function has never supported (and does not today) TF >2.0.0.b. 
Their build is simply run with old PIP. 

We have spent over week of dev time trying and failing, and then another week trying to get whats wrong through paid support channels, because we have believed the docs & marketing... just like you.. 
That blog is just false (its using 2.0.0b while claiming 2.0), documentation is false (referring to 2.0 when its 2.0.0b) and even tech support was false for first 3 days (until we have documented everything and only thing left was to acknowledge that they have a problem).

Beyond that, similar problems can be found in their other products claiming TF support, its either old Python, old pip or size restrictions...

**Comment by northtree** [comment]
@m4recek That's not true. 

[2.0.0](https://github.com/tensorflow/tensorflow/releases/tag/v2.0.0) was released on Oct 2019 which is newer than [2.0.0b0](https://github.com/tensorflow/tensorflow/releases/tag/v2.0.0-beta0). 

Currently, I could deploy GCF with `tensorflow==2.0.0` in requirements.txt. 

@bonaventura-p You are right. TF 2.1 is not supported and will return `Build Error` on deployment.

**Comment by m4recek** [comment]
@northtree I think you are missing the timeline of the original issue.
- June 7, 2019 - 2.0.0b0 beta release
- July 9, 2019 - Blog is published, presumably docs updated (thats why its using 2.0.0b0)
- Sept 30, 2019 - 2.0.0 stable release
- Nov 29, 2019 - this thread, public & private bug reports created (google cloud functions were running pip 19)
- April 2, 2020 - last comment in public tracker asking for help https://issuetracker.google.com/issues/145374253

Which means Blog, Marketing & Doc were all wrong... for 10 months?
Its great if 2.0 was finally fixed in last few weeks. I guess the marketing announcement didn't happened this time and bug tickets were not updated. Also TF is on 2.1, almost 2.2 now..

It seems that TF is iterating too quickly for GCP to keep up 😄 

Off topic: We have tried Google's AI platform with similarly mixed results in terms of docs & marketing accuracy. Right now it seems like the best way to run TF models in production apps is in custom build containers in Cloud Run or Kubernetes.
